rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,299 @@
+- Start Date: 2019-05-11
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Handlers are asynchronous functions registered with core services invoked to
+respond to events like a HTTP request, or mounting an application. _Handler
+context_ is a pattern that would allow APIs and values to be provided to handler
+functions by the service that owns the handler (aka service owner) or other
+services that are not necessarily known to the service owner.
+
+# Basic example
+
+```js
+// services can register context providers to route handlers
+http.registerContext('myApi', request => ({ getId() { return request.params.myApiId } }));
```

Comment:
Does `registerContext` extends a context globally? I think would be useful to let users to declare plugin specific extension point.
```js
const options = { private: true };
http.router.registerRequestContext('privateApi', async request => {
  return myPrivateApi.do(request)
}, options);
```

Reply:
I'm going to push back on this since there are other ways of giving handlers access to plugin-specific objects. This pattern is intended to solve the problem of providing _exposed_ services to other plugins, when the service owner doesn't have control of the consuming code (the handler).

In the case of plugins wanting to share objects within their own handlers, they can easily do so via closures:

```ts
class Plugin {
  start(core) {
    const myPluginService = new GraphQlClient();

    core.http.router.route({
      path: '/api/foo',
      method: 'POST',
      handler(context, request) {
        return myPluginService.query();
      }
    });
  }
}
```

Reply:
agree, I told the same `plugin may pass dependencies explicitly`.
although it's creates the same problems that the current proposal solves:
- needs to track service updates manually.
```js
class Plugin {
  start(core) {
    const myPluginService$ = new GraphQlClient();

    core.http.router.route({
      path: '/api/foo',
      method: 'POST',
      handler(context, request) {
        const myPluginService = await myPluginService$.pipe(first()).toPromise();
        return myPluginService.query();
     }
   });
 }
}
```
- needs to configure api manually
```js
class Plugin {
  start(core) {
    const myPluginService$ = new GraphQlClient();

    core.http.router.route({
      path: '/api/foo',
      method: 'POST',
      handler(context, request) {
        const myPluginService = await myPluginService$.pipe(first()).toPromise();
        return myPluginService.asScoped(request).query();
     }
   });
 }
}
```

I don't insist we should implement it, just noted it could be useful in some circumstances.

Reply:
If I'm following correctly, there are two ways to achieve this:

Option 1:
```
// plugin_a/plugin.ts
class PluginA {
  start(core) {
   const myPluginService$ = new GraphQlClient();

   async function getClient(request){
      const myPluginService = await myPluginService$.pipe(first()).toPromise();
      return myPluginService.asScoped(request)
   }
  
  http.registerContext('getClient', getClient);

   return {}
  }
}

// plugin_b/plugin.ts
class PluginB {
  start(core, plugins) {

   core.http.router.route({
      path: '/api/foo',
      method: 'POST',
      handler(context, request) {
        return await context.pluginA.getClient(request).mutate(...);
      }
   });
  }
}
```

Option 2:
```
// plugin_a/plugin.ts
class PluginA {
  start(core) {
   const myPluginService$ = new GraphQlClient();

   async function getClient(request){
      const myPluginService = await myPluginService$.pipe(first()).toPromise();
      return myPluginService.asScoped(request)
   }

   return {
     getClient
   }
  }
}

// plugin_b/plugin.ts
class PluginB {
  start(core, plugins) {

   core.http.router.route({
      path: '/api/foo',
      method: 'POST',
      handler(context, request) {
        return await plugins.pluginA.getClient(request).mutate(...);
      }
   });
  }
}
```

and this RFC is encouraging the former.  Which of the five key problems does the former solve but not the latter?

Reply:
Maybe that is covered in this chunk:

> The obvious alternative is what we've always done: expose all functionality at
the plugin level and then leave it up to the consumer to build a "context" for
their particular handler. This creates a lot of inconsistency and makes
creating simple but useful handlers more complicated. This can also lead to
subtle but significant bugs as it's unreasonable to assume all developers
understand the important details for constructing a context with plugins they
don't know anything about.

which I guess makes sense, it's about simplifying what's available on `context` so it only contains things that are relevant.